test/e2e/reboot.go

```diff
@@ -64,52 +64,60 @@ var _ = Describe("Reboot", func() {
 	})
 
 	It("should reboot each node and ensure they function upon restart", func() {
-		// This test requires SSH, so the provider check should be identical to
-		// there (the limiting factor is the implementation of util.go's
-		// getSigner(...)).
-		provider := testContext.Provider
-		if !providerIs("gce", "gke") {
-			By(fmt.Sprintf("Skipping reboot test, which is not implemented for %s", provider))
-			return
-		}
-
-		// Get all nodes, and kick off the test on each.
-		nodelist, err := c.Nodes().List(labels.Everything(), fields.Everything())
-		if err != nil {
-			Failf("Error getting nodes: %v", err)
-		}
-		result := make(chan bool, len(nodelist.Items))
-		for _, n := range nodelist.Items {
-			go rebootNode(c, provider, n.ObjectMeta.Name, result)
-		}
+		testReboot(c, true)
+	})
 
-		// Wait for all to finish and check the final result.
-		failed := false
-		// TODO(mbforbes): Change to `for range` syntax and remove logging once
-		// we support only Go >= 1.4.
-		for _, n := range nodelist.Items {
-			if !<-result {
-				Failf("Node %s failed reboot test.", n.ObjectMeta.Name)
-				failed = true
-			}
-		}
-		if failed {
-			Failf("Test failed; at least one node failed to reboot in the time given.")
-		}
+	It("should cause kernel panic on each node and ensure they function upon restart", func() {
+		testReboot(c, false)
 	})
 })
 
+func testReboot(c *client.Client, clean bool) {
```

Comment:
I think that this could actually be made much more general and useful with some relatively minor changes.  All this function really currently does is enumerate all nodes, call a function against each one in a go routine, and check that all of the function calls succeeded.  That seems like a useful generalization in itself.  There might even be an implementation of it in pkg/utils already - worth checking.  If not, perhaps the function to be called could simply be passed in as a closure?

Secondly, regarding the boolean second argument, I think that there are at least three cases here, and possibly more in future.  This is mostly a comment about the rebootNode function below. You already have two options (and a boolean flag to choose between these):

1. sudo reboot /* clean shutdown and restart */
2. echo c | sudo tee /proc/sysrq-trigger /* kernel panic */

I think we also need something like:

3. ifdown eth0 && sleep 120 && ifup eth0 /* switch the network interface off for a while to simulate a network outage */
4. iptables -A INPUT -j DROP && sleep 120 && iptables -D INPUT -j DROP /* tell the firewall to drop all inbound packets for a while */
4. iptables -A OUTPUT -j DROP && sleep 120 && iptables -D OUTPUT -j DROP /* tell the firewall to drop all outbound packets for a while */

... and probably a few more over time.

Hence I suggest generalizing this, by either passing the command string in as a parameter, or changing the second boolean parameter to a enumeration.